marimo/_server/sessions.py

```diff
@@ -124,24 +134,86 @@ def launch_kernel_with_cleanup(*args: Any) -> None:
             self.kernel_task = threading.Thread(
                 target=launch_kernel_with_cleanup,
                 args=(
-                    self.control_queue,
-                    self.input_queue,
+                    self.queue_manager.control_queue,
+                    self.queue_manager.input_queue,
                     listener.address,
                     is_edit_mode,
-                    configs,
+                    self.configs,
                 ),
                 # daemon threads can create child processes, unlike
                 # daemon processes
                 daemon=True,
             )
+
         self.kernel_task.start()
         # First thing kernel does is connect to the socket, so it's safe to
         # call accept
-        self.read_conn = listener.accept()
+        self._read_conn = listener.accept()
+
+    def is_alive(self) -> bool:
+        return self.kernel_task is not None and self.kernel_task.is_alive()
+
+    def interrupt_kernel(self) -> None:
+        if isinstance(self.kernel_task, mp.Process) and self.kernel_task.pid:
+            LOGGER.debug("Sending SIGINT to kernel")
+            os.kill(self.kernel_task.pid, signal.SIGINT)
+
+    def close_kernel(self) -> None:
+        assert self.kernel_task is not None, "kernel not started"
+
+        if isinstance(self.kernel_task, mp.Process):
+            self.queue_manager.close_queues()
+            if self.kernel_task.is_alive():
+                self.kernel_task.terminate()
+            self.kernel_connection.close()
+        elif self.kernel_task.is_alive():
+            self.queue_manager.control_queue.put(requests.StopRequest())
+            self.kernel_task.join()
```

Comment:
this was the only addition. no code other code paths should have changed, but my tests would not pass unless i joined here